core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java

```diff
@@ -0,0 +1,777 @@
+            }
+            @Override
+            public Long getTo(Range r) {
+                return (long)r.to - (r.includeTo ? 0 : 1);
+            }
+            @Override
+            public Field getRangeField(String name, Range r) {
+                return new LongRangeField(name, new long[] {getFrom(r)}, new long[] {getTo(r)});
+            }
+            @Override
+            public byte[] getBytes(Range r) {
+                byte[] b = new byte[Long.BYTES*2];
+                long from = r.from == null ? Long.MIN_VALUE : getFrom(r);
+                long to = r.to == null ? Long.MAX_VALUE : getTo(r);
+                NumericUtils.longToSortableBytes(from, b, 0);
+                NumericUtils.longToSortableBytes(to, b, Long.BYTES);
+                return b;
+            }
+            @Override
+            public Query withinQuery(String field, Number from, Number to, boolean includeFrom, boolean includeTo) {
+                return LongRangeField.newWithinQuery(field,  new long[] {(Long)from + (includeFrom ? 0 : 1)},
+                    new long[] {(Long)to - (includeTo ? 0 : 1)});
+            }
+            @Override
+            public Query containsQuery(String field, Number from, Number to, boolean includeFrom, boolean includeTo) {
+                return LongRangeField.newContainsQuery(field,  new long[] {(Long)from + (includeFrom ? 0 : 1)},
+                    new long[] {(Long)to - (includeTo ? 0 : 1)});
+            }
+            @Override
+            public Query intersectsQuery(String field, Number from, Number to, boolean includeFrom, boolean includeTo) {
+                return LongRangeField.newIntersectsQuery(field,  new long[] {(Long)from + (includeFrom ? 0 : 1)},
+                    new long[] {(Long)to - (includeTo ? 0 : 1)});
+            }
+        };
+
+        RangeType(String name, NumberType type) {
+            this.name = name;
+            this.numberType = type;
+        }
+
+        /** Get the associated type name. */
+        public final String typeName() {
+            return name;
+        }
+
+        protected abstract byte[] getBytes(Range range);
+        public abstract Field getRangeField(String name, Range range);
+        public List<IndexableField> createFields(String name, Range range, boolean indexed, boolean docValued, boolean stored) {
+            assert range != null : "range cannot be null when creating fields";
+            List<IndexableField> fields = new ArrayList<>();
+            if (indexed) {
+                fields.add(getRangeField(name, range));
+            }
+            // todo add docValues ranges once aggregations are supported
+            if (stored) {
+                fields.add(new StoredField(name, range.toString()));
+            }
+            return fields;
+        }
+        public Number parse(RangeFieldType fieldType, XContentParser parser, boolean coerce, LongSupplier nowInMillis) throws IOException {
```

Comment:
this method seems to be only used at indexing time, so I don't think it should accept `nowInMillis` since index dates need to be concrete dates rather than math expressions